// ebos_public/public/js/ebos_globalize.js
(function( root, factory ) {

	// UMD returnExports
	if ( typeof define === "function" && define.amd ) {

		// AMD
		define([
			"cldr",
			"../globalize"
		], factory );
	} else if ( typeof exports === "object" ) {

		// Node, CommonJS
		module.exports = factory( require('./globalize/globalize'), require("./globalize/message") );
	} else {
		// Extend global
		factory( root.Globalize );
		Globalize.locale(root.$('#ebos').attr('lang'));
	}
}(this, function( Globalize ) {
	Globalize.load({
		"main": {
			"en": {
				"identity": {
			    	"version": {
			      		"_cldrVersion": "26",
	    		  		"_number": "$Revision: 10969 $"
	    			},
	    			"generation": {
	      				"_date": "$Date: 2014-09-11 12:17:53 -0500 (Thu, 11 Sep 2014) $"
	    			},
	    			"language": "en"
	    		},
	    		"numbers": {
	    			"defaultNumberingSystem": "latn",
	    			"otherNumberingSystems": {
	    				"native": "latn"
	    			},
	    			"minimumGroupingDigits": "1",
	    			"symbols-numberSystem-latn": {
	    				"decimal": ".",
	    				"group": ",",
	    				"list": ";",
	    				"percentSign": "%",
	    				"plusSign": "+",
	    				"minusSign": "-",
	    				"exponential": "E",
	    				"superscriptingExponent": "×",
	    				"perMille": "‰",
	    				"infinity": "∞",
	    				"nan": "NaN",
	    				"timeSeparator": ":"
	    			},
	    			"decimalFormats-numberSystem-latn": {
	    				"standard": "#,##0.###",
	    				"long": {
	    					"decimalFormat": {
	    						"1000-count-one": "0 thousand",
	    						"1000-count-other": "0 thousand",
	    						"10000-count-one": "00 thousand",
	    						"10000-count-other": "00 thousand",
	    						"100000-count-one": "000 thousand",
	    						"100000-count-other": "000 thousand",
	    						"1000000-count-one": "0 million",
	    						"1000000-count-other": "0 million",
	    						"10000000-count-one": "00 million",
	    						"10000000-count-other": "00 million",
	    						"100000000-count-one": "000 million",
	    						"100000000-count-other": "000 million",
	    						"1000000000-count-one": "0 billion",
	    						"1000000000-count-other": "0 billion",
	    						"10000000000-count-one": "00 billion",
	    						"10000000000-count-other": "00 billion",
	    						"100000000000-count-one": "000 billion",
	    						"100000000000-count-other": "000 billion",
	    						"1000000000000-count-one": "0 trillion",
	    						"1000000000000-count-other": "0 trillion",
	    						"10000000000000-count-one": "00 trillion",
	    						"10000000000000-count-other": "00 trillion",
	    						"100000000000000-count-one": "000 trillion",
	    						"100000000000000-count-other": "000 trillion"
	    					}
	    				},
	    				"short": {
	    					"decimalFormat": {
	    						"1000-count-one": "0K",
	    						"1000-count-other": "0K",
	    						"10000-count-one": "00K",
	    						"10000-count-other": "00K",
	    						"100000-count-one": "000K",
	    						"100000-count-other": "000K",
	    						"1000000-count-one": "0M",
	    						"1000000-count-other": "0M",
	    						"10000000-count-one": "00M",
	    						"10000000-count-other": "00M",
	    						"100000000-count-one": "000M",
	    						"100000000-count-other": "000M",
	    						"1000000000-count-one": "0B",
	    						"1000000000-count-other": "0B",
	    						"10000000000-count-one": "00B",
	    						"10000000000-count-other": "00B",
	    						"100000000000-count-one": "000B",
	    						"100000000000-count-other": "000B",
	    						"1000000000000-count-one": "0T",
	    						"1000000000000-count-other": "0T",
	    						"10000000000000-count-one": "00T",
	    						"10000000000000-count-other": "00T",
	    						"100000000000000-count-one": "000T",
	    						"100000000000000-count-other": "000T"
	    					}
	    				}
	    			},
	    			"scientificFormats-numberSystem-latn": {
	    				"standard": "#E0"
	    			},
	    			"percentFormats-numberSystem-latn": {
	    				"standard": "#,##0%"
	    			},
	    			"currencyFormats-numberSystem-latn": {
	    				"currencySpacing": {
	    					"beforeCurrency": {
	    						"currencyMatch": "[:^S:]",
	    						"surroundingMatch": "[:digit:]",
	    						"insertBetween": " "
	    					},
	    					"afterCurrency": {
	    						"currencyMatch": "[:^S:]",
	    						"surroundingMatch": "[:digit:]",
	    						"insertBetween": " "
	    					}
	    				},
	    				"accounting": "¤#,##0.00;(¤#,##0.00)",
	    				"standard": "¤#,##0.00",
	    				"unitPattern-count-one": "{0} {1}",
	    				"unitPattern-count-other": "{0} {1}"
	    			},
	    			"miscPatterns-numberSystem-latn": {
	    				"atLeast": "{0}+",
	    				"range": "{0}–{1}"
	    			}
	    		}
			},
			"es": {
				"identity": {
			    	"version": {
			      		"_cldrVersion": "26",
	    		  		"_number": "$Revision: 10969 $"
	    			},
	    			"generation": {
	      				"_date": "$Date: 2014-09-11 12:17:53 -0500 (Thu, 11 Sep 2014) $"
	    			},
	    			"language": "es"
	    		},
				"numbers": {
					"defaultNumberingSystem": "latn",
					"otherNumberingSystems": {
						"native": "latn"
					},
					"minimumGroupingDigits": "1",
					"symbols-numberSystem-latn": {
						"decimal": ",",
						"group": ".",
						"list": ";",
						"percentSign": "%",
						"plusSign": "+",
						"minusSign": "-",
						"exponential": "E",
						"superscriptingExponent": "×",
						"perMille": "‰",
						"infinity": "∞",
						"nan": "NaN",
						"timeSeparator": ":"
					},
					"decimalFormats-numberSystem-latn": {
						"standard": "#,##0.###",
						"long": {
							"decimalFormat": {
								"1000-count-one": "0 mil",
								"1000-count-other": "0 mil",
								"10000-count-one": "00 mil",
								"10000-count-other": "00 mil",
								"100000-count-one": "000 mil",
								"100000-count-other": "000 mil",
								"1000000-count-one": "0 millón",
								"1000000-count-other": "0 millones",
								"10000000-count-one": "00 millones",
								"10000000-count-other": "00 millones",
								"100000000-count-one": "000 millones",
								"100000000-count-other": "000 millones",
								"1000000000-count-one": "0 mil millones",
								"1000000000-count-other": "0 mil millones",
								"10000000000-count-one": "00 mil millones",
								"10000000000-count-other": "00 mil millones",
								"100000000000-count-one": "000 mil millones",
								"100000000000-count-other": "000 mil millones",
								"1000000000000-count-one": "0 billón",
								"1000000000000-count-other": "0 billones",
								"10000000000000-count-one": "00 billones",
								"10000000000000-count-other": "00 billones",
								"100000000000000-count-one": "000 billones",
								"100000000000000-count-other": "000 billones"
							}
						},
						"short": {
							"decimalFormat": {
								"1000-count-one": "0K",
								"1000-count-other": "0K",
								"10000-count-one": "00K",
								"10000-count-other": "00K",
								"100000-count-one": "000K",
								"100000-count-other": "000K",
								"1000000-count-one": "0M",
								"1000000-count-other": "0M",
								"10000000-count-one": "00M",
								"10000000-count-other": "00M",
								"100000000-count-one": "000M",
								"100000000-count-other": "000M",
								"1000000000-count-one": "0000M",
								"1000000000-count-other": "0000M",
								"10000000000-count-one": "00MRD",
								"10000000000-count-other": "00MRD",
								"100000000000-count-one": "000MRD",
								"100000000000-count-other": "000MRD",
								"1000000000000-count-one": "0B",
								"1000000000000-count-other": "0B",
								"10000000000000-count-one": "00B",
								"10000000000000-count-other": "00B",
								"100000000000000-count-one": "000B",
								"100000000000000-count-other": "000B"
							}
						}
					},
					"scientificFormats-numberSystem-latn": {
						"standard": "#E0"
					},
					"percentFormats-numberSystem-latn": {
						"standard": "#,##0 %"
					},
					"currencyFormats-numberSystem-latn": {
						"currencySpacing": {
							"beforeCurrency": {
								"currencyMatch": "[:^S:]",
								"surroundingMatch": "[:digit:]",
								"insertBetween": " "
							},
							"afterCurrency": {
								"currencyMatch": "[:^S:]",
								"surroundingMatch": "[:digit:]",
								"insertBetween": " "
							}
						},
						"accounting": "#,##0.00 ¤",
						"standard": "#,##0.00 ¤",
						"unitPattern-count-one": "{0} {1}",
						"unitPattern-count-other": "{0} {1}"
					},
					"miscPatterns-numberSystem-latn": {
						"atLeast": "Más de {0}",
						"range": "{0}-{1}"
					}
				}
			},
			"fr": {
				"identity": {
			    	"version": {
			      		"_cldrVersion": "26",
	    		  		"_number": "$Revision: 10969 $"
	    			},
	    			"generation": {
	      				"_date": "$Date: 2014-09-11 12:17:53 -0500 (Thu, 11 Sep 2014) $"
	    			},
	    			"language": "fr"
	    		},
	    		"numbers": {
	    			"defaultNumberingSystem": "latn",
	    			"otherNumberingSystems": {
	    				"native": "latn"
	    			},
	    			"minimumGroupingDigits": "1",
	    			"symbols-numberSystem-latn": {
	    				"decimal": ",",
	    				"group": " ",
	    				"list": ";",
	    				"percentSign": "%",
	    				"plusSign": "+",
	    				"minusSign": "-",
	    				"exponential": "E",
	    				"superscriptingExponent": "×",
	    				"perMille": "‰",
	    				"infinity": "∞",
	    				"nan": "NaN",
	    				"timeSeparator": ":"
	    			},
	    			"decimalFormats-numberSystem-latn": {
	    				"standard": "#,##0.###",
	    				"long": {
	    					"decimalFormat": {
	    						"1000-count-one": "0 millier",
	    						"1000-count-other": "0 mille",
	    						"10000-count-one": "00 millier",
	    						"10000-count-other": "00 mille",
	    						"100000-count-one": "000 millier",
	    						"100000-count-other": "000 mille",
	    						"1000000-count-one": "0 million",
	    						"1000000-count-other": "0 millions",
	    						"10000000-count-one": "00 million",
	    						"10000000-count-other": "00 millions",
	    						"100000000-count-one": "000 million",
	    						"100000000-count-other": "000 millions",
	    						"1000000000-count-one": "0 milliard",
	    						"1000000000-count-other": "0 milliards",
	    						"10000000000-count-one": "00 milliard",
	    						"10000000000-count-other": "00 milliards",
	    						"100000000000-count-one": "000 milliards",
	    						"100000000000-count-other": "000 milliards",
	    						"1000000000000-count-one": "0 billion",
	    						"1000000000000-count-other": "0 billions",
	    						"10000000000000-count-one": "00 billions",
	    						"10000000000000-count-other": "00 billions",
	    						"100000000000000-count-one": "000 billions",
	    						"100000000000000-count-other": "000 billions"
	    					}
	    				},
	    				"short": {
	    					"decimalFormat": {
	    						"1000-count-one": "0 k",
	    						"1000-count-other": "0 k",
	    						"10000-count-one": "00 k",
	    						"10000-count-other": "00 k",
	    						"100000-count-one": "000 k",
	    						"100000-count-other": "000 k",
	    						"1000000-count-one": "0 M",
	    						"1000000-count-other": "0 M",
	    						"10000000-count-one": "00 M",
	    						"10000000-count-other": "00 M",
	    						"100000000-count-one": "000 M",
	    						"100000000-count-other": "000 M",
	    						"1000000000-count-one": "0 Md",
	    						"1000000000-count-other": "0 Md",
	    						"10000000000-count-one": "00 Md",
	    						"10000000000-count-other": "00 Md",
	    						"100000000000-count-one": "000 Md",
	    						"100000000000-count-other": "000 Md",
	    						"1000000000000-count-one": "0 Bn",
	    						"1000000000000-count-other": "0 Bn",
	    						"10000000000000-count-one": "00 Bn",
	    						"10000000000000-count-other": "00 Bn",
	    						"100000000000000-count-one": "000 Bn",
	    						"100000000000000-count-other": "000 Bn"
	    					}
	    				}
	    			},
	    			"scientificFormats-numberSystem-latn": {
	    				"standard": "#E0"
	    			},
	    			"percentFormats-numberSystem-latn": {
	    				"standard": "#,##0 %"
	    			},
	    			"currencyFormats-numberSystem-latn": {
	    				"currencySpacing": {
	    					"beforeCurrency": {
	    						"currencyMatch": "[:^S:]",
	    						"surroundingMatch": "[:digit:]",
	    						"insertBetween": " "
	    					},
	    					"afterCurrency": {
	    						"currencyMatch": "[:^S:]",
	    						"surroundingMatch": "[:digit:]",
	    						"insertBetween": " "
	    					}
	    				},
	    				"accounting": "#,##0.00 ¤;(#,##0.00 ¤)",
	    				"standard": "#,##0.00 ¤",
	    				"unitPattern-count-one": "{0} {1}",
	    				"unitPattern-count-other": "{0} {1}"
	    			},
	    			"miscPatterns-numberSystem-latn": {
	    				"atLeast": "au moins {0}",
	    				"range": "{0}–{1}"
	    			}
	    		}
	    	}
		},
	  	"supplemental": {
	    	"version": {
	      		"_cldrVersion": "26",
	      		"_number": "$Revision: 10969 $"
	    		},
	    	"generation": {
	      		"_date": "$Date: 2014-09-11 12:17:53 -0500 (Thu, 11 Sep 2014) $"
	    	},
	    	"likelySubtags": {
	      		"en": "en-Latn-US",
	      		"es": "es-Latn-ES",
	    	  	"fr": "fr-Latn-FR"
	    	}
	  	}
	});

	Globalize.loadMessages({
		en: {
			error: "Error",
			fiveDotTwo: "Values must be numeric with up to 2 decimals. (Example: 12345.99)",
			fiveDotTwoRegex: "/^[0-9]{0,5}(\.[0-9]{0,2})?$/",
			integer: "Values must be numeric with no decimals. (Example: 123)",
			previousInv: "Previous Inventories for Store Number:",
			openInv: "Click a row to open inventory:",
			date: "Date",
			time: "Time",
			lastSaved: "Last Saved",
			createdBy: "Created By",
			status: "Status",
			listType: "List Type",
			orPrint: "OR Select a blank inventory to print:",
			startNewInv: "Start New Inventory",
			selectFreq: "Select Frequency Type",
			selectDateTime: "Select a Date and Time for your new inventory:",
			cancel: "Cancel",
			startInventory: "Start Inventory",
			alreadyInv: "There is already an inventory on ",
			atLower: " at ",
			likeToEdit: "Would you like to edit that inventory",
			likeToView: "Would you like to view that inventory",
			editInv: "Edit Inventory",
			viewInv: "View Inventory",
			Daily: "Daily",
			Weekly: "Weekly",
			Monthly: "Monthly",
			"In-Progress": "In-Progress",
			Completed: "Completed",
			lastSavedAt: "Last Saved At",
			lastSavedBy: "Last Saved By",
			neverSaved: "Never Saved",
			description: "Description",
			outerPack: "Outer Pack",
			innerPack: "Inner Pack",
			looseUnit: "Loose Unit",
			totalUnits: "Total Units",
			search: "Search",
			post: "Post",
			noLongerEdit: "You will no longer be able to edit this inventory.",
			aboutToChange: "You are about to change your list from ",
			toSpaces: " to ",
			onSpaces: " on ",
			atSpaces: " at ",
			areSure: "Are You Sure?",
			invSaved: "Your Inventory has been saved.",
			success: "Success",
			timeStamp: "Time Stamp",
			user: "User",
			fieldName: "Field Name",
			beforeValue: "Before Value",
			afterValue: "After Value",
			case_count: "Outer Packs",
			inner_packcount: "Inner Packs",
			loose_count: "Loose Units",
			units_count: "Total Units",
			date_created: "Last Saved",
			basis_list: "List Type",
			auditFor: "Audit For ",
			zeroRecords: "No search results found.",
			Dash_1: "Dashboard Item One",
			Dash_2: "Dashboard Item Two",
			Dash_3: "Dashboard Item Three",
			Dash_4: "Dashboard Item Four",
			notice: "Notice",
			audit: "Audit",
			nsn: "National Store Number",
			register_id: "Register ID",
			profitability: "eProfitability",
			inventory: "Inventory",
			log_monitor: "Log Monitor",
			store:	"Store",
			physinv: "Physical Inventory",
			drawer_countdown: "Drawer Countdown",
			drawer_count: "Drawer Count",
			drawer_banks: "Drawer Banks",
			save: "Save",
			cashAmt: "Cash Amount",
			click_to_display: "Click a row to display details",
			store_busn_dt: "Store Business Date",
			cashout_time: "Cashout Time",
			crew_name: "Crew Name",
			expect_cash: "Expected Cash",
			status: "Status",
			cash_over_short: "Cash Over/Short",
			current_forever: "Current Forever",
			close_date_time: "Close Date & Time",
			register: "Register",
			preparer: "Preparer",
			shift_mgr: "Shift Manager",
			cashier: "Cashier",
			forever_diff: "Forever Difference",
			pos_overrings: "POS overrings",
			cash_refunds: "Cash Refunds",
			man_ref_overrngs: "Manual Refunds/Overrings",
			billable_sales: "Billable Sales",
			non_taxable_sales: "Non Taxable Sales",
			cashless_sales: "Cashless Sales",
			gift_certs_redm_in_dols: "Gift Certificates Redeemed (in $)",
			arch_cards_redm_in_dols: "Arch Card Redeemed (in $)",
			skims: "Skims",
			exp_drwr_cash: "Expected Drawer Cash",
			drwr_count_cash: "Drawer Count Cash",
			trans_count: "Transaction Count",
			avg_check: "Average Check",
			one: "One",
			two: "Two",
			three: "Three",
			four: "Four",
			five: "Five",
			six: "Six",
			seven: "Seven",
			eight: "Eight",
			nine: "Nine",
			ten: "Ten",
			eleven: "Eleven",
			twelve: "Twelve",
			thirteen: "Thirteen",
			fourteen: "Fourteen",
			fifteen: "Fifteen",
			sixteen: "Sixteen",
			seventeen: "Seventeen",
			eighteen: "Eighteen",
			nineteen: "Nineteen",
			twenty: "Twenty",
			dc_business_date: "Drawer Countdown Business Date",
			sc_business_date: "Safe Count Business Date",
			env_business_date: "Envelope Business Date",
			dc_detail:"Drawer Count Detail",
			register: "Register",
			cashout_timestamp : "Cashout Timestamp",
			shift_manager: "Shift Manager",
			cashier: "Cashier",
			manual_refs_overrings: "Manual Refund/Overrings",
			billable_sales: "Billable Sales",
			non_taxable_sales: "Non Taxable Sales",
			gift_certificate_redeemed: "Gift Cert. Redeemed",
			dc_cash: "Drawer Count Cash",
			promotion : "Promo",
			discount_coupon: "Discount/Coupon",
			deposit_code: "Deposit Code",
			preparer: "Preparer",
			drawer_bank: "Drawer Bank",
			forever_diff: "Forever Difference",
			pos_overrings: "POS Overrings",
			cash_refunds: "Cash Refunds",
			cashless_sales: "Cashless Sales",
			arch_card_redeemed: "Arch Card Redeemed",
			expected_drawer_cash: "Expected Drawer Cash",
			cash_over_short: "Cash Over/Short",
			transaction_count: "Transaction Count",
			average_check: "Average Check",
			cash: "Cash",
			expected: "Expected",
			actual: "Actual",
			over_short: "Over/Short",
			discount_amount: "Discount/Amount",
			before_tred: "Before-TRED",
			after_tred: "After-TRED",
			manager: "Manager",
			sales: "Sales",
			close_time: "Close Time",
			type: "Type",
			no_cashout_found: "No registers have been cashed out for the selected business date:",
			currency_calculator: "CURRENCY CALCULATOR",
			coin: "Coin",
			roll: "Roll",
			value: "Value",
			or: "or",
			coin_total: "Coin Total",
			cash_to_leave: "Cash to leave in drawer",
			bill: "Bill",
			count: "Count",
			total: "Total",
			travelers_checks: "Travelers Checks",
			commit: "Commit",
			foreign_currency: "Foreign Currency",
			change_from_coins: "Change from Coins",
			total_counted_cash: "Total Counted Cash",
			safe_count: "Safe Count",
			close: "Close",
			store_control_settings : "Store Control Settings",
			safe_settings : "Safe Settings",
			safe_amount: "Safe Amount",
			petty_cash_amount : "Petty Cash Amount",
			enable_gift_certificates : "Gift Certificates",
			enable_other : "Other",
			till_quantity: "Till Quantity",
			coin_magazine_quantity: "Coin Magazine Quantity",
			envelope_settings : "Envelope Settings",
			enable_envelopes : "Enable",
			id_required: "Id Required",
			envelope_id_val: "Envelope identifier can only be two characters - either alpha, numeric, or a combination of. Examples: 12, AB, 1A, 2B, etc.",
			name: "Name",
			envelope_id: "Envelope ID",
			other: "Other",
			seettingsSaved: "Your Settings have been saved.",
			purchases_for_store_number: "Purchases for Store Number",
			pending_purchases: "Pending Purchases",
			enter_manual_purchase: "Enter Manual Purchase",
			view_store_ledger: "View Store Ledger",
			purchase_history_list: "Purchase History List",
			start_date: "Start Date",
			end_date: "End Date",
			vendor: "Vendor",
			all_vendors: "ALL VENDORS",
			get_purchase_history: "Get Purchase History",
			posted_purchase_detail: "Posted Purchase Detail",
			invoice_id: "Invoice ID",
			vendor_name: "Vendor Name",
			delivery_date: "Delivery Date",
			amount: "Amount",
			tax: "Tax",
			state_tax: "State Tax",
			manual_invoice: "Manual Invoice",
			edit: "Edit",
			new: "New",
			invoice_date: "Invoice Date",
			invoice_number: "Invoice Number",
			wrin: "WRIN",
			uom: "UOM",
			uom_per_case: "UOM/Case",
			case: "Case",
			quantity: "Quantity",
			subtotal: "Sub-total",
			remove: "Remove",
			unsaved_data: "Unsaved Data",
			unsaved_data_close_wo_save: "There is unsaved data. Are you sure you want to Close without saving",
			ok: "OK",
			unable_to_find_vendor: "Unable to find selected vendor",
			unable_to_delete_invoice: "Unable to delete invoice",
			invoice_posted: "Your invoice has been posted",
			invoice_saved: "Your invoice has been saved",
			enter_valid_value: "Please enter a valid value for the highlighted field(s). Fields must have a value and be greater than 0, otherwise the row should be removed from the invoice",
			invoice_requires_one_line: "An invoice requires at least one line item",
			pending_purchase_details: "Pending Purchase Details",
			line: "LINE",
			case_price: "Case Price",
			cases_purchased: "Cases Purchased",
			cost: "Cost",
			purchase: "Purchase",
			store_ledger_for: "Store Ledger for Store Number",
			select_store_ledger_month: "Select Store Ledger by Month",
			save_as_csv: "Save as CSV",
			store_ledger: "store_ledger",
			add: "Add",
			no_items_found: "No items were found for the selected vendor",
			records_found: "Records Found",
			safe_calculator: "Safe Calculator",
			mag_selector: "Magazine Selector",
			calculator: "Calculator",
			magazine: "Magazine",
			petty_cash: "Petty Cash",
			gift_cert_short: "G.C.",
			no_records_found: "No Records Found",
			till: "Till",
			safe: "Safe",
			user_id: "User ID",
			enable_coin_magazines : "Coin Magazines",
			coin_magazine_amount : "Coin Magazine Amount",
			envelope_amount: "Envelope Amount",
			discount: "Discount",
			coupon: "Coupon",
			emp_meal_disc: "Employee Meal Discount",
			mgr_meal_disc: "Manager Meal Discount",
			promo_dol_pct: "Promo dollars as a % of sales from the drawer",
			promo_pct_exceeds_var: "!!! PROMO PERCENT EXCEEDS VARIANCE !!!",
			allow_promo_var: "Allowable Promo Variance",
			disc_dol_pct: "Discount dollars as a % of sales from the drawer",
			disc_pct_exceeds_var: "!!! DISCOUNT PERCENT EXCEEDS VARIANCE !!!",
			allow_disc_var: "Allowable Discount Variance",
			coupon_dol_pct: "Coupon dollars as a % of sales from the drawer",
			coupon_pct_exceeds_var: "!!! COUPON PERCENT EXCEEDS VARIANCE !!!",
			allow_coupon_var: "Allowable Coupon Variance",
			edit_drawer_banks: "Edit Drawer Banks",
			other_name: "Other Name",
			variance_settings: "Variance Settings",
			alpha_deposit_cd: "Alpha Deposit Code",
			daypart_deposit_cd: "Daypart Deposit Code",
			multiple_deposits: "Enable Multiple Deposits",
			Changed: "Changed",
			Closed: "Closed",
			Counted: "Counted",
			Error: "Error",
			Inprogress: "In-Progress",
			InProgress: "In-Progress",
			New: "New",
			NotSupported: "Not Supported",
			Pending: "Pending",
			Posted: "Posted",
			Processed: "Processed",
			Uncounted: "Uncounted",
			Voided: "Voided",
			deposit_validation: "Deposit Validation",
			envelopes: "Envelopes",
			deposits: "Deposits",
			crime_losses: "Crime Losses",
			skim_business_date: "Skim Business Date",
			skim_type: "Skim Type",
			preparer_id: "Preparer ID",
			preparer_name: "Preparer Name",
			skim_timestamp: "Skim Timestamp",
			created_timestamp: "Created Timestamp",
			skim_amount: "Skim Amount",
			register_number: "Register #:",
			entry_time: "Entry Time:",
			manager_colon: "Manager:",
			skim_source_colon: "Skim Source:",
			deposit_code_colon: "Deposit Code:",
			envelope_id_colon: "Envelope ID:",
			skim_amount_colon: "Skim Amount:",
			Morning: "Morning",
			Lunch: "Lunch",
			dinner: "Dinner",
			cancel: "Cancel",
			save: "Save",
			finalize: "Finalize",
			Finalize: "Finalize",
			manual: "Manual",
			Manual: "Manual",
			POS: "POS",
			add_new_skim: "Add New Skim",
			skims_audit: "Skims Audit",
			error_finding_skim_record: "Error retrieving audit records. Please try again later.",
			scheduled: "Scheduled",
			unscheduled: "Unscheduled",
			foreign_currency: "Foreign Currency",
			amount_usd: "Amount (USD)",
			open: "Open",
			closed: "Closed",
			of_sales: "of Sales",
			coupon_a: "Coupon A",
			coupon_b: "Coupon B",
			coupon_c: "Coupon C",
			coupon_d: "Coupon D",
			coupon_e: "Coupon E",
			foreign_currency_in_usd: "Foreign Currency(in USD)",
			other_receipts: "Other Receipts",
			employee_meals: "Employee Meals",
			manager_meals: "Manager Meals",
			counted_cash: "Counted Cash",
			source: "Source",
			bag_num: "Bag Number",
			counted_deposit: "Counted Deposit",
			depositor_name: "Depositor Name",
			error: "Error",
			fiveDotTwo: "Values must be numeric with up to 2 decimals. (Example: 12345.99)",
			fiveDotTwoRegex: "/^[0-9]{0,5}(\.[0-9]{0,2})?$/",
			integer: "Values must be numeric with no decimals. (Example: 123)",
			previousInv: "Previous Inventories for Store Number:",
			openInv: "Click a row to open inventory:",
			date: "Date",
			time: "Time",
			lastSaved: "Last Saved",
			createdBy: "Created By",
			status: "Status",
			listType: "List Type",
			orPrint: "OR Select a blank inventory to print:",
			startNewInv: "Start New Inventory",
			selectFreq: "Select Frequency Type",
			selectDateTime: "Select a Date and Time for your new inventory:",
			cancel: "Cancel",
			startInventory: "Start Inventory",
			alreadyInv: "There is already an inventory on ",
			atLower: " at ",
			likeToEdit: "Would you like to edit that inventory",
			likeToView: "Would you like to view that inventory",
			editInv: "Edit Inventory",
			viewInv: "View Inventory",
			Daily: "Daily",
			Weekly: "Weekly",
			Monthly: "Monthly",
			"In-Progress": "In-Progress",
			Completed: "Completed",
			lastSavedAt: "Last Saved At",
			lastSavedBy: "Last Saved By",
			neverSaved: "Never Saved",
			description: "Description",
			outerPack: "Outer Pack",
			innerPack: "Inner Pack",
			looseUnit: "Loose Unit",
			totalUnits: "Total Units",
			search: "Search",
			post: "Post",
			noLongerEdit: "You will no longer be able to edit this inventory.",
			aboutToChange: "You are about to change your list from ",
			toSpaces: " to ",
			onSpaces: " on ",
			atSpaces: " at ",
			areSure: "Are You Sure?",
			invSaved: "Your Inventory has been saved.",
			success: "Success",
			timeStamp: "Time Stamp",
			user: "User",
			fieldName: "Field Name",
			beforeValue: "Before Value",
			afterValue: "After Value",
			case_count: "Outer Packs",
			inner_packcount: "Inner Packs",
			loose_count: "Loose Units",
			units_count: "Total Units",
			date_created: "Last Saved",
			basis_list: "List Type",
			auditFor: "Audit For ",
			zeroRecords: "No search results found.",
			Dash_1: "Dashboard Item One",
			Dash_2: "Dashboard Item Two",
			Dash_3: "Dashboard Item Three",
			Dash_4: "Dashboard Item Four",
			notice: "Notice",
			audit: "Audit",
			nsn: "National Store Number",
			register_id: "Register ID",
			profitability: "eProfitability",
			inventory: "Inventory",
			log_monitor: "Log Monitor",
			store:	"Store",
			physinv: "Physical Inventory",
			drawer_countdown: "Drawer Countdown",
			drawer_count: "Drawer Count",
			drawer_banks: "Drawer Banks",
			save: "Save",
			cashAmt: "Cash Amount",
			click_to_display: "Click a row to display details",
			store_busn_dt: "Store Business Date",
			cashout_time: "Cashout Time",
			crew_name: "Crew Name",
			expect_cash: "Expected Cash",
			status: "Status",
			cash_over_short: "Cash Over/Short",
			current_forever: "Current Forever",
			close_date_time: "Close Date & Time",
			register: "Register",
			preparer: "Preparer",
			shift_mgr: "Shift Manager",
			cashier: "Cashier",
			forever_diff: "Forever Difference",
			pos_overrings: "POS overrings",
			cash_refunds: "Cash Refunds",
			man_ref_overrngs: "Manual Refunds/Overrings",
			billable_sales: "Billable Sales",
			non_taxable_sales: "Non Taxable Sales",
			tax_id_num: "Tax ID Number",
			cashless_sales: "Cashless Sales",
			gift_certs_redm_in_dols: "Gift Certificates Redeemed (in $)",
			arch_cards_redm_in_dols: "Arch Card Redeemed (in $)",
			skims: "Skims",
			exp_drwr_cash: "Expected Drawer Cash",
			drwr_count_cash: "Drawer Count Cash",
			trans_count: "Transaction Count",
			avg_check: "Average Check",
			one: "One",
			two: "Two",
			three: "Three",
			four: "Four",
			five: "Five",
			six: "Six",
			seven: "Seven",
			eight: "Eight",
			nine: "Nine",
			ten: "Ten",
			eleven: "Eleven",
			twelve: "Twelve",
			thirteen: "Thirteen",
			fourteen: "Fourteen",
			fifteen: "Fifteen",
			sixteen: "Sixteen",
			seventeen: "Seventeen",
			eighteen: "Eighteen",
			nineteen: "Nineteen",
			twenty: "Twenty",
			dc_business_date: "Drawer Countdown Business Date",
			sc_business_date: "Safe Count Business Date",
			env_business_date: "Envelope Business Date",
			dc_detail:"Drawer Count Detail",
			register: "Register",
			cashout_timestamp : "Cashout Timestamp",
			shift_manager: "Shift Manager",
			cashier: "Cashier",
			manual_refs_overrings: "Manual Refund/Overrings",
			billable_sales: "Billable Sales",
			non_taxable_sales: "Non Taxable Sales",
			gift_certificate_redeemed: "Gift Cert. Redeemed",
			dc_cash: "Drawer Count Cash",
			promotion : "Promo",
			discount_coupon: "Discount/Coupon",
			deposit_code: "Deposit Code",
			preparer: "Preparer",
			drawer_bank: "Drawer Bank",
			forever_diff: "Forever Difference",
			pos_overrings: "POS Overrings",
			cash_refunds: "Cash Refunds",
			cashless_sales: "Cashless Sales",
			arch_card_redeemed: "Arch Card Redeemed",
			expected_drawer_cash: "Expected Drawer Cash",
			cash_over_short: "Cash Over/Short",
			transaction_count: "Transaction Count",
			average_check: "Average Check",
			cash: "Cash",
			expected: "Expected",
			actual: "Actual",
			over_short: "Over/Short",
			discount_amount: "Discount/Amount",
			before_tred: "Before-TRED",
			after_tred: "After-TRED",
			manager: "Manager",
			sales: "Sales",
			close_time: "Close Time",
			type: "Type",
			no_cashout_found: "No registers have been cashed out for the selected business date:",
			currency_calculator: "CURRENCY CALCULATOR",
			coin: "Coin",
			roll: "Roll",
			value: "Value",
			or: "or",
			coin_total: "Coin Total",
			cash_to_leave: "Cash to leave in drawer",
			bill: "Bill",
			count: "Count",
			total: "Total",
			travelers_checks: "Travelers Checks",
			commit: "Commit",
			foreign_currency: "Foreign Currency",
			change_from_coins: "Change from Coins",
			total_counted_cash: "Total Counted Cash",
			safe_count: "Safe Count",
			close: "Close",
			store_control_settings : "Store Control Settings",
			safe_settings : "Safe Settings",
			safe_amount: "Safe Amount",
			petty_cash_amount : "Petty Cash Amount",
			enable_gift_certificates : "Gift Certificates",
			enable_other : "Other",
			till_quantity: "Till Quantity",
			coin_magazine_quantity: "Coin Magazine Quantity",
			envelope_settings : "Envelope Settings",
			enable_envelopes : "Enable",
			id_required: "Id Required",
			envelope_id_val: "Envelope identifier can only be two characters - either alpha, numeric, or a combination of. Examples: 12, AB, 1A, 2B, etc.",
			name: "Name",
			envelope_id: "Envelope ID",
			other: "Other",
			seettingsSaved: "Your Settings have been saved.",
			purchases_for_store_number: "Purchases for Store Number",
			pending_purchases: "Pending Purchases",
			enter_manual_purchase: "Enter Manual Purchase",
			view_store_ledger: "View Store Ledger",
			purchase_history_list: "Purchase History List",
			start_date: "Start Date",
			end_date: "End Date",
			vendor: "Vendor",
			all_vendors: "ALL VENDORS",
			get_purchase_history: "Get Purchase History",
			posted_purchase_detail: "Posted Purchase Detail",
			invoice_id: "Invoice ID",
			vendor_name: "Vendor Name",
			delivery_date: "Delivery Date",
			amount: "Amount",
			tax: "Tax",
			state_tax: "State Tax",
			manual_invoice: "Manual Invoice",
			edit: "Edit",
			new: "New",
			invoice_date: "Invoice Date",
			invoice_number: "Invoice Number",
			wrin: "WRIN",
			uom: "UOM",
			uom_per_case: "UOM/Case",
			case: "Case",
			quantity: "Quantity",
			subtotal: "Sub-total",
			remove: "Remove",
			unsaved_data: "Unsaved Data",
			unsaved_data_close_wo_save: "There is unsaved data. Are you sure you want to Close without saving",
			ok: "OK",
			unable_to_find_vendor: "Unable to find selected vendor",
			unable_to_delete_invoice: "Unable to delete invoice",
			invoice_posted: "Your invoice has been posted",
			invoice_saved: "Your invoice has been saved",
			enter_valid_value: "Please enter a valid value for the highlighted field(s). Fields must have a value and be greater than 0, otherwise the row should be removed from the invoice",
			invoice_requires_one_line: "An invoice requires at least one line item",
			pending_purchase_details: "Pending Purchase Details",
			line: "LINE",
			case_price: "Case Price",
			cases_purchased: "Cases Purchased",
			cost: "Cost",
			purchase: "Purchase",
			store_ledger_for: "Store Ledger for Store Number",
			select_store_ledger_month: "Select Store Ledger by Month",
			save_as_csv: "Save as CSV",
			store_ledger: "store_ledger",
			add: "Add",
			no_items_found: "No items were found for the selected vendor",
			records_found: "Records Found",
			safe_calculator: "Safe Calculator",
			mag_selector: "Magazine Selector",
			calculator: "Calculator",
			magazine: "Magazine",
			petty_cash: "Petty Cash",
			gift_cert_short: "G.C.",
			no_records_found: "No Records Found",
			till: "Till",
			safe: "Safe",
			user_id: "User ID",
			enable_coin_magazines : "Coin Magazines",
			coin_magazine_amount : "Coin Magazine Amount",
			envelope_amount: "Envelope Amount",
			discount: "Discount",
			coupon: "Coupon",
			emp_meal_disc: "Employee Meal Discount",
			mgr_meal_disc: "Manager Meal Discount",
			promo_dol_pct: "Promo dollars as a % of sales from the drawer",
			promo_pct_exceeds_var: "!!! PROMO PERCENT EXCEEDS VARIANCE !!!",
			allow_promo_var: "Allowable Promo Variance",
			disc_dol_pct: "Discount dollars as a % of sales from the drawer",
			disc_pct_exceeds_var: "!!! DISCOUNT PERCENT EXCEEDS VARIANCE !!!",
			allow_disc_var: "Allowable Discount Variance",
			coupon_dol_pct: "Coupon dollars as a % of sales from the drawer",
			coupon_pct_exceeds_var: "!!! COUPON PERCENT EXCEEDS VARIANCE !!!",
			allow_coupon_var: "Allowable Coupon Variance",
			edit_drawer_banks: "Edit Drawer Banks",
			other_name: "Other Name",
			variance_settings: "Variance Settings",
			alpha_deposit_cd: "Alpha Deposit Code",
			daypart_deposit_cd: "Daypart Deposit Code",
			multiple_deposits: "Enable Multiple Deposits",
			Changed: "Changed",
			Closed: "Closed",
			Counted: "Counted",
			Error: "Error",
			Inprogress: "In-Progress",
			InProgress: "In-Progress",
			New: "New",
			NotSupported: "Not Supported",
			Pending: "Pending",
			Posted: "Posted",
			Processed: "Processed",
			Uncounted: "Uncounted",
			Voided: "Voided",
			deposit_validation: "Deposit Validation",
			envelopes: "Envelopes",
			deposits: "Deposits",
			crime_losses: "Crime Losses",
			skim_business_date: "Skim Business Date",
			skim_type: "Skim Type",
			preparer_id: "Preparer ID",
			preparer_name: "Preparer Name",
			skim_timestamp: "Skim Timestamp",
			created_timestamp: "Created Timestamp",
			skim_amount: "Skim Amount",
			register_number: "Register #:",
			entry_time: "Entry Time:",
			manager_colon: "Manager:",
			skim_source_colon: "Skim Source:",
			deposit_code_colon: "Deposit Code:",
			envelope_id_colon: "Envelope ID:",
			skim_amount_colon: "Skim Amount:",
			Morning: "Morning",
			Lunch: "Lunch",
			dinner: "Dinner",
			cancel: "Cancel",
			save: "Save",
			finalized: "Finalized",
			Finalized: "Finalized",
			manual: "Manual",
			Manual: "Manual",
			POS: "POS",
			add_new_skim: "Add New Skim",
			skims_audit: "Skims Audit",
			error_finding_skim_record: "Error retrieving audit records. Please try again later.",
			scheduled: "Scheduled",
			unscheduled: "Unscheduled",
			foreign_currency: "Foreign Currency",
			amount_usd: "Amount (USD)",
			open: "Open",
			closed: "Closed",
			of_sales: "of Sales",
			coupon_a: "Coupon A",
			coupon_b: "Coupon B",
			coupon_c: "Coupon C",
			coupon_d: "Coupon D",
			coupon_e: "Coupon E",
			foreign_currency_in_usd: "Foreign Currency(in USD)",
			other_reciepts: "Other Receipts",
			employee_meals: "Employee Meals",
			manager_meals: "Manager Meals",
			counted_cash: "Counted Cash",
			source: "Source",
			bag_num: "Bag Number",
			counted_deposit: "Counted Deposit",
			depositor_name: "Depositor Name",
			attribute: "Attribute",
			long_desc: "Long Description",
			prod_cat: "Product Category",
			uom_per_sleeve: "UOM/Sleeve",
			inner_pack_desc: "Inner Pack Description",
			temp_zone: "Temperature Zone",
			mcd_gl_acct: "McDonalds GL Account",
			prim_vendor: "Primary Vendor",
			cur_use_per_1000: "Current Usage/$1000",
			use_per_1000_hist: "Usage Per $1000/History",
			cost_per_unit: "Cost per Unit",
			done_with_item: "Done With This Item",
			enter_wrin_or_desc: "Enter Raw Item WRIN or Description",
			raw_item_info: "View and Edit Raw Item Information",
			no_app_items_found: "No Applicable Items Found",
			changes_saved: "Changes Saved",
			projected_cash: "Projected Deposit Amount",
			deposit_validation_business_date: "Deposit Validation Date",
			depositor: "Depositor",
			validator: "Validator",
			deposit_amount: "Deposit Amount",
			validated_amount: "Validated Amount",
			expected_amount: "Expected Amount",
			no_deposits_to_validate: "No Deposits to Validate",
			manual_refs_overrings_colon: "Manual Refunds and Overrings:",
			gift_certificate_redeemed_colon: "Gift Certificates Redeemed:",
			dc_cash_colon: "Drawer Counted Cash:",
			petty_cash_desc: "Petty Cash Description",
			validated: "Validated",
			Validated: "Validated",
			complete: "Complete",
			reason_deposit_validation_does_not_match: "Validated amount does not match expected amount please enter a reason if known:",
			reason_validation_does_not_match: "Reason the expected amount does not match the validated amount:",
			Skim: "Skim",
			DrawerCount: "Drawer Count",
			no_contributors: "This deposit does not have any skims or drawer counts at this time.",
			attribute: "Attribute",
			long_desc: "Long Description",
			prod_cat: "Product Category",
			uom_per_sleeve: "UOM/Sleeve",
			inner_pack_desc: "Inner Pack Description",
			temp_zone: "Temperature Zone",
			mcd_gl_acct: "McDonalds GL Account",
			prim_vendor: "Primary Vendor",
			cur_use_per_1000: "Current Usage/$1000",
			use_per_1000_hist: "Usage Per $1000/History",
			avgd_cost_per_unit: "Averaged Cost per Unit",
			done_with_item: "Done With This Item",
			enter_wrin_or_desc: "Enter Raw Item WRIN or Description",
			raw_item_info: "Raw Item Information",
			no_app_items_found: "No Applicable Items Found",
			changes_saved: "Changes Saved",
			raw_item_activity: "Raw Item Activity",
			select_date_range: "Select Date Range",
			get_item_details: "Get Item Details",
			addl_info: "Additional Information",
			system_date_time: "System Date and Time",
			user_date_time: "User Entered Date and Time",
			event: "Event",
			view_details: "View Details",
			messages: "Messages",
			contributors: "Contributors",
			Complete: "Complete",
			delete: "Delete",
			created_time: "Created Time",
			show_crime_loss: "Show Crime Loss",
			cl_loss_time: "Loss Occurrence Time",
			cl_loss_date: "Loss Occurrence Date",
			cl_loss_amount: "Loss Amount",
			cl_loss_description: "Loss Description",
			cl_loss_discovery_timestamp: "Loss Discovery Date",
			cl_loss_police_agency: "Police Agency",
			cl_loss_police_report_id: "Police Report ID",
			cl_loss_injuries: "Were there injuries?",
			cl_loss_injury_desc: "Injury Description",
			cl_loss_physical_damage: "Was there physical damage?",
			cl_loss_damage_desc: "Physical Damage Description",
			cl_insurance_claim_id: "Insurance Claim ID",
			"true":"True",
			"True":"True",
			"False":"False",
			"false":"False",
			crime_loss:"Crime Loss",
			rm_crime_loss: "Remove Crime Loss",
			cloud: "Cloud",
			local: "Local",
			unable_to_save_deposit: "Unable to Save Deposit",
			deposit_totals_differ: "Projected deposit totals are different than counted deposit total. Please return to the Skims and/or Drawer Countdown screens to balance counted cash with projected deposit amount.",
			counted: "Counted",
			projected: "Projected",
			difference: "Difference",
			local: "Local",
			cloud: "Cloud",
			skim: "Skim",
			manual_vendors: "Manual Vendors",
			add_drawer_bank: "Add Drawer Bank",
			new_register_number: "New Register Number:",
			deposit_codes: "Deposit Codes",
            date_of_loss: "Date of Loss",
            time_of_loss: "Time of Loss",
            manager_on_duty: "Manager on Duty",
            loss_discovered: "Loss Discovered",
            police_report: "Police Report",
            loss_amount: "Loss Amount",
            showing_losses_between: "Showing losses between",
            crime_loss_date_range: "Crime Loss Summary Range",
            "90days": "90 days",
            "6months": "6 months",
            "9months": "9 months",
            "12months": "12 months",
            enterValidTransferQty: "Please enter the quantity of the item you are transferring.",
            Loss_Occurrence_Date_Cannot_be_Greater_than_the_Discovery_Date: "Loss Occurrence Date Cannot be Greater than the Discovery Date.",
            Loss_Occurrence_Date_set_to_Discovery_Date: "Loss Occurrence Date has been set to Discovery Date.",
            warning: "Warning",
            deposit_contributors: "Deposit Contributors",
            transfer_incomplete: "Warning: Transfer Incomplete",
            warn_transfer_will_be_lost: "All entered information will be lost.  Are you sure you want to cancel?",
            continue_transfer: "Continue Transfer",
            cancel_transfer: "Cancel Transfer",
            export_csv: "Export CSV",
            manager_id: "Manager ID",
            police_agency: "Police Agency",
            loss_description: "Loss Description",
            loss_injuries: "Loss Injuries",
            loss_injuries_desc: "Injury Description",
            physical_damage: "Physical Damage",
            physical_damage_desc: "Physical Damage Description",
            insurance_claim_id: "Insurance Claim ID",
            updated_timestamp: "Updated Timestamp",
            bag_number: "Bag Number",
            projected_deposit_cash: "Projected Deposit Cash",
            deposit_cash: "Deposit Cash",
            verified_deposit_cash: "Verified Deposit Cash",
            verified_bank_date:"Verified Bank Date",
            verified_reason: "Verified Reason",
            verifier: "Verifier",
            verifier_name: "Verifier Name",
            start_date: "Start Date",
            "empty": "",
            retrieve_records: "Retrieve Records",
            no_crime_losses: "No Crime Losses",
            dates_reset: "The end date cannnot be lower than the start date. Dates Reset.",
            custom_date_range: "Custom Date Range",
            other_amount: "Other Amount",
            gift_cert: "Gift Certificate",
            update_timestamp: "Update Timestamp",
			pending_crime_loss_warning: "Cannot finalize deposit until pending crime loss is handled",
            Pending_cl: "Crime Loss Pending Approval",
			Finalized_cl: "Crime Loss Approved",
			Rejected_cl: "Crime Loss Rejected",
			promo_waste: "Promotions & Waste",
			raw: "Raw",
			compl_promo_waste: "Completed Promo & Waste",
			raw_item: "Raw Item",
			entered: "Entered",
			outer: "Outer",
			inner: "Inner",
			loose: "Loose",
			promotion_long: "Promotion",
			waste: "Waste",
			begin_date: "Begin Date",
			set_begin_date: "Set Beginning Date",
			month_to_date: "Month to Date",
			two_month_to_date: "Two Months to Date",
			daily_rollup: "Daily Rollup",
			click_to_view_or_edit: "Click a row to view details or edit.",
			add_new_entry: "Add New Entry",
			items: "Items",
			click_plus_sign: "Click Plus Sign to Add Item",
			invoice_type: "Invoice Type",
			invoice_amt: "Invoice Amount",
			food: "Food",
			paper: "Paper",
			ops_supplies: "Ops Supplies",
			linens: "Linens",
			non_prod_hpy_ml: "Non-product: Happy Meal",
			non_prod_other: "Non-product: Other",
			invoice_time: "Invoice Time",
			enter_quick_search: "Enter Quick Search With Suggestions for Manual Purchases",
			optional: "Optional",
			dol_case_var_off: "Invoice total dollar amount and case quantity total exceed the variances calculated by eInventory.",
			dol_var_off: "This invoice total dollar amount exceeds the variance calculated by eInventory.",
			case_var_off: "This invoice case quantity total exceeds the variance calculated by eInventory.",
			call_dist_center: "Call your distribution center.",
			del_invoice: "Delete Invoice",
			post_var_off_inv: "Post Variance Off Invoice",
			enter_wrin_or_desc: "Enter Raw Item WRIN or Description",
			when_did_truck_leave: "When did your truck pull off the parking lot?",
			organization_name: "Organization Name",
			phone_number: "Phone Number",
			email: "E-Mail",
			address: "Address",
			invoice_or_po: "Invoice or PO Number",
			user_name: "User Name",
			billable_sales_date_range: "Billable Sales Date Range",
			city: "City",
			state: "State",
			zipcode: "Zipcode",
			address_line_1: "Address Line 1",
			address_line_2: "Address Line 2",
			contact_name: "Contact Name",
			phoneNumberRegex: "Only Numbers up to Ten Places Allowed",
			other_receipts: "Other Receipts",
			food: "Food",
			paper: "Paper",
			other: "Other",
			organization_name: "Organization Name",
			contact_name: "Contact Name",
			check_number: "Check Number",
			address_line_1: "Address Line 1",
			address_line_2: "Address Line 2",
        },
		es: {
			error: "Error",
			fiveDotTwo: "Los valores deben ser numéricos con hasta 2 decimales (Ejemplo: 12345,99)",
			fiveDotTwoRegex: "/^[0-9]{0,5}(\,[0-9]{0,2})?$/",
			integer: "Los valores deben ser numérico sin decimales. (Ejemplo: 123)",
			previousInv: "Inventarios previos para Número tienda:",
			openInv: "Haga clic en una fila para abrir el inventario:",
			date: "Fecha",
			time: "Tiempo",
			lastSaved: "Guardó Por Ultima Vez",
			createdBy: "Preado Por",
			status: "Estado",
			listType: "Tipo de Lista",
			orPrint: "O Seleccione un inventario en blanco para imprimir:",
			startNewInv: "Iniciar un Inventario Nuevo",
			selectFreq: "Seleccionar Tipo de Frecuencia",
			selectDateTime: "Seleccione una Fecha y Hora para su nuevo inventario:",
			cancel: "Cancelar",
			startInventory: "Iniciar Inventario",
			alreadyInv: "Ya existe un inventario de ",
			atLower: " en ",
			likeToEdit: "¿Te gustaría editar ese inventario?",
			likeToView: "¿Le gustaría ver que el inventario?",
			editInv: "Editar Inventario",
			viewInv: "Ver Inventario",
			Daily: "Diario",
			Weekly: "Semanal",
			Monthly: "Mensual",
			"In-Progress": "En Curso",
			Completed: "Terminado",
			lastSavedAt: "Guardado por última vez en",
			lastSavedBy: "Guardado por",
			neverSaved: "Nunca Guardados",
			description: "Descripción",
			outerPack: "Paquete Exterior",
			innerPack: "Paquete Interior",
			looseUnit: "Unidad Loose",
			totalUnits: "Unidades Totales",
			search: "Búsqueda",
			post: "Puesto",
			save: "Guardar",
			noLongerEdit: "Usted ya no será capaz de editar este inventario.",
			aboutToChange: "Usted está a punto de cambiar su lista de ",
			toSpaces: " a ",
			onSpaces: " en ",
			atSpaces: " en ",
			areSure: "Estás seguro?",
			invSaved: "Tu inventario se ha guardado.",
			success: "Éxito",
			timeStamp: "Marca de Tiempo",
			user: "Usuario",
			fieldName: "Nombre del Campo",
			beforeValue: "Antes de Valor",
			afterValue: "Después de Valor",
			case_count: "Paquete Exterior",
			inner_packcount: "Paquete Interior",
			loose_count: "Unidad Loose",
			units_count: "Unidades Totales",
			date_created: "Guardó por Ultima Vez",
			basis_list: "Tipo de Lista",
			status: "Estado",
			auditFor: "Auditoría para ",
			zeroRecords: "No hay resultados encontrados.",
			Dash_1: "Dashboard Punto Uno",
			Dash_2: "Dashboard Punto Segundo",
			Dash_3: "Dashboard Punto Tercero",
			Dash_4: "Dashboard Punto Cuarto",
			notice: "Aviso",
			audit: "Auditoría",
			nsn: "Nacional Número Tienda",
			register_id: "Registro ID",
			profitability: "eRentabilidad",
			inventory: "Inventario",
			cash: "Efectivo",
			log_monitor: "Log Monitor",
			store:	"Restaurante",
			physinv: "Físico Inventario",
			drawer_countdown: "Cajón Cuenta Atrás",
			drawer_count: "Cajón Contar",
			drawer_banks: "Cajón Banks",
			save: "Guardar",
			cashAmt: "Efectivo Monto",
			click_to_display: "Haga clic en una fila para mostrar los detalles",
			store_busn_dt: "Tienda Negocios Fecha",
			cashout_time: "Cashout Tiempo",
			crew_name: "Tripulación Nombre",
			expect_cash: "Efectivo Esperado",
			status: "Estado",
			cash_over_short: "Efectivo Encima/Corto",
			current_forever: "Corriente para Siempre",
			close_date_time: "Cerca de la Fecha & Hora",
			register: "Registro",
			preparer: "Preparador",
			shift_mgr: "Turno Gerente",
			cashier: "Cajero",
			forever_diff: "Siempre Diferencia",
			pos_overrings: "POS Encima Costo",
			cash_refunds: "Efectivo Reembolsos",
			man_ref_overrngs: "Manual Reembolsos/Encima Costo",
			billable_sales: "Facturable Venta",
			non_taxable_sales: "Non Taxable Sales",
			cashless_sales: "Sin Dinero en Efectivo Venta",
			gift_certs_redm_in_dols: "Certificado de Regalo Redimido (en $)",
			arch_cards_redm_in_dols: "Arch Tarjeta Redimido (en $)",
			skims: "Descremadas",
			exp_drwr_cash: "Efectivo Cajón Esperado",
			drwr_count_cash: "Cajón Contar Efectivo",
			trans_count: "Transacción Contar",
			avg_check: "Promedio Cheque",
			one: "Uno",
			two: "Dos",
			three: "Tres",
			four: "Quatro",
			five: "Cinco",
			six: "Seis",
			seven: "Siete",
			eight: "Ocho",
			nine: "Nueve",
			ten: "Diez",
			eleven: "Once",
			twelve: "Doce",
			thirteen: "Trece",
			fourteen: "Catorce",
			fifteen: "Quince",
			sixteen: "Dieciséis",
			seventeen: "Diecisiete",
			eighteen: "Dieciocho",
			nineteen: "Diecinueve",
			twenty: "Veinte",
			dc_business_date: "Cajón Cuenta atrás Negocios Fecha",
			sc_business_date: "Seguro Cuenta atrás Negocios Fecha",
			env_business_date: "Sobre de la Fecha de Negocios",
			dc_detail:"Detalle del Conde Cajón",
			register: "Registro",
			cashout_timestamp : "Marca de tiempo del cambio cajón",
			shift_manager: "Gerente de turno de trabajo",
			cashier: "Cajero",
			manual_refs_overrings: "Manual de Reembolso / Overrings",
			billable_sales: "Ventas facturables",
			non_taxable_sales: "Non Taxable Sales",
			gift_certificate_redeemed: "Vale de regalo redimió",
			dc_cash: "Cash contó en el cajón",
			promotion : "Promoción",
			discount_coupon: "Descuento/Cupón",
			deposit_code: "Código de Depósito",
			preparer: "Preparador",
			drawer_bank: "Cajón Banco",
			current_forever: "Actual para Siempre",
			forever_diff: "Siempre Diferencia",
			pos_overrings: "POS Overrings(sp)",
			cash_refunds: "Reembolsos en Efectivo",
			cashless_sales: "Sin Efectivo Ventas",
			arch_card_redeemed: "Arco Tarjeta Redimidos",
			skims: "Reducción",
			expected_drawer_cash: "Esperado Cajón del Efectivo",
			transaction_count: "Conde de Transacción",
			average_check: "Tarifa promedio",
			cash: "Efectivo",
			expected: "Esperado",
			actual: "Real",
			over_short: "Mas/Bajo",
			discount_amount: "Descuento/Monto",
			before_tred: "Antes-TRED",
			after_tred: "Después-TRED",
			manager: "Gerente",
			sales: "Venta",
			close_time: "Cerrar tiempo",
			type: "Tipo",
			status: "Estado",
			no_cashout_found: "No hay registros se han cobrado a cabo para la fecha seleccionada negocio:",
			currency_calculator: "CALCULADORA DE DIVISAS",
			coin: "Esquina",
			roll: "Rollo",
			value: "Valor",
			or: "o",
			coin_total: "Total de Moneda",
			cash_to_leave: "Dinero en efectivo para dejar en el cajón",
			bill: "Billete",
			count: "Contar",
			total: "Total",
			travelers_checks: "Cheques de Viaje",
			commit: "Cometer",
			foreign_currency: "Moneda Extranjera",
			change_from_coins: "Cambio de monedas",
			total_counted_cash: "Total de efectivo contado",
			safe_count: "Recuento Seguro",
			close: "Cerrar",
            user_id: "User ID",
            enable_coin_magazines : "Coin Magazines",
            coin_magazine_amount : "Coin Magazine Amount",
            envelope_amount: "Envelope Amount"

        },
		fr: {
			error: "Erreur",
			fiveDotTwo: "Les valeurs doivent être numériques avec jusqu'à 2 décimales. (Exemple: 12345,99)",
			fiveDotTwoRegex: "/^[0-9]{0,5}(\,[0-9]{0,2})?$/",
			integer: "Les valeurs doivent être numériques sans décimales. (Exemple: 123)",
			previousInv: "Inventaires précédents pour le numéro de magasin:",
			openInv: "Cliquez sur une ligne pour ouvrir l'inventaire:",
			date: "Date",
			time: "Temps",
			lastSaved: "Dernier Enregistrement",
			createdBy: "Créé Par",
			status: "Statut",
			listType: "Type de Liste",
			orPrint: "OU Sélectionnez un inventaire vierge à imprimer:",
			startNewInv: "Start New Inventory",
			selectFreq: "Select Frequency Type",
			selectDateTime: "Select a Date and Time for your new inventory:",
			cancel: "Cancel",
			startInventory: "Start Inventory",
			alreadyInv: "There is already an inventory on ",
			atLower: " at ",
			likeToEdit: "Would you like to edit that inventory",
			likeToView: "Would you like to view that inventory",
			editInv: "Edit Inventory",
			viewInv: "View Inventory",
			Daily: "Daily",
			Weekly: "Weekly",
			Monthly: "Monthly",
			"In-Progress": "In-Progress",
			Completed: "Completed",
			lastSavedAt: "Guardado por última vez en",
			lastSavedBy: "Guardado por",
			neverSaved: "Nunca Guardados",
			description: "Description",
			outerPack: "Outer Pack",
			innerPack: "Inner Pack",
			looseUnit: "Loose Unit",
			totalUnits: "Total Units",
			search: "Search",
			post: "Post",
			save: "Save",
			noLongerEdit: "You will no longer be able to edit this inventory.",
			aboutToChange: "You are about to change your list from ",
			toSpaces: " to ",
			onSpaces: " on ",
			atSpaces: " at ",
			areSure: "Are You Sure?",
			invSaved: "Your Inventory has been saved.",
			success: "Success",
			timeStamp: "Time Stamp",
			user: "User",
			fieldName: "Field Name",
			beforeValue: "Before Value",
			afterValue: "After Value",
			case_count: "Outer Packs",
			inner_packcount: "Inner Packs",
			loose_count: "Loose Units",
			units_count: "Total Units",
			date_created: "Last Saved",
			basis_list: "List Type",
			status: "Status",
			auditFor: "Audit For ",
			zeroRecords: "No search results found.",
			Dash_1: "Dashboard Point One",
			Dash_2: "Dashboard Point Deux",
			Dash_3: "Dashboard Point Trois",
			Dash_4: "Dashboard Point Quatre",
			notice: "Avis",
			audit: "Audit",
			nsn: "National Nombre Magasin",
			register_id: "Caisse ID",
			profitability: "eRentabilité",
			inventory: "Inventaire",
			cash: "Espèces",
			log_monitor: "Log Monitor",
			store:	"Restaurant",
			physinv: "Physique Inventaire",
			drawer_countdown: "Tiroir Compte à Rebours",
			drawer_count: "Tiroir Compter",
			drawer_banks: "Tiroir Banks",
			save: "Sauver",
			cashAmt: "Montant en Espèces",
			click_to_display: "Cliquez sur une ligne pour afficher les détails",
			store_busn_dt: "Magasin Indiquant la Date d'Affaires",
			cashout_time: "le Temps d'Encaissement",
			crew_name: "Nom Equipage",
			expect_cash: "Attendu Espèces",
			status: "Statut",
			cash_over_short: "Espèces Sur/Court",
			current_forever: "Courant Toujours",
			close_date_time: "Fini Date & l'Heure",
			register: "Caisse",
			preparer: "Préparateur",
			shift_mgr: "Equipe Gérant",
			cashier: "Caissier",
			forever_diff: "Toujours Différence",
			pos_overrings: "POS Surcharge",
			cash_refunds: "Espèces Remboursements",
			man_ref_overrngs: "Manuel Remboursements/Surcharge",
			billable_sales: "Facturable Ventes",
			non_taxable_sales: "Non Taxable Sales",
			cashless_sales: "Sans Numéraire Ventes",
			gift_certs_redm_in_dols: "Cadeau Certificats Rachetés (en $)",
			arch_cards_redm_in_dols: "Arch Carte Rachetés (en $)",
			skims: "Ecrémé",
			exp_drwr_cash: "Attendu Tiroir Espèces",
			drwr_count_cash: "Tiroir Compter Espèces",
			trans_count: "Transaction Compter",
			avg_check: "Moyenne Chèque",
			one: "Une",
			two: "Deux",
			three: "Trois",
			four: "Quatre",
			five: "Cinq",
			six: "Six",
			seven: "Sept",
			eight: "Huit",
			nine: "Nuef",
			ten: "Dix",
			eleven: "Onze",
			twelve: "Douze",
			thirteen: "Treize",
			fourteen: "Quatorze",
			fifteen: "Quinze",
			sixteen: "Seize",
			seventeen: "Dix-sept",
			eighteen: "Dix-huit",
			nineteen: "Dix-neuf",
			twenty: "Vingt",
			dc_business_date: "Tiroir Countdown affaires Date",
			sc_business_date: "Comte sécurité affaires Date",
			env_business_date: "Cnveloppe Date d'entreprise",
			dc_detail:"Tiroir comte Détail",
			register: "Se enregistrer",
			cashout_timestamp : "Cashout Horodatage",
			shift_manager: "Gestionnaire de Quart",
			cashier: "Caissier",
			manual_refs_overrings: "Manuel Remboursement/Overrings",
			billable_sales: "Les ventes facturables",
			non_taxable_sales: "Non Taxable Sales",
			gift_certificate_redeemed: "Certificat cadeau Racheté",
			dc_cash: "Nombre de tiroir-caisse",
			promotion : "Promo",
			discount_coupon: "Remise/Coupon",
			deposit_code: "Code de Dépôt",
			preparer: "Préparateur",
			drawer_bank: "Tiroir Banque",
			current_forever: "Toujours actuel",
			forever_diff: "Toujours Différence",
			pos_overrings: "POS Overrings(fr)",
			cash_refunds: "Les remboursements en espèces",
			cashless_sales: "Scripturaux Ventes",
			arch_card_redeemed: "Arche carte Racheté",
			skims: "Réduction",
			expected_drawer_cash: "Prévue tiroir-caisse",
			cash_over_short: "Espèces supérieurs / Short",
			transaction_count: "Nombre de transactions",
			average_check: "Vérification de la moyenne",
			cash: "Espèces",
			expected: "Attendu",
			actual: "Réel",
			over_short: "Over/Court",
			discount_amount: "Remise/Montant",
			before_tred: "Avant-TRED",
			after_tred: "Après-TRED",
			manager: "Directeur",
			sales: "Ventes",
			close_time: "Heure de Clôture",
			type: "Type",
			status: "Statut",
			no_cashout_found: "Aucun registres ont été encaissés pour la date de visite sélectionnée:",
			currency_calculator: "CONVERTISSEUR DE DEVISES",
			coin: "Coin",
			roll: "Rouleau",
			value: "Valeur",
			or: "ou",
			coin_total: "Pièce Totale",
			cash_to_leave: "Trésorerie de laisser dans le tiroir",
			bill: "Billet",
			count: "Compter",
			total: "Total",
			travelers_checks: "Chèques de Voyage",
			commit: "Commettre",
			foreign_currency: "Devises étrangères",
			change_from_coins: "Changement de pièces de monnaie",
			total_counted_cash: "Comptant Total Compté",
			safe_count: "Comte Sécurité",
			close: "Fermer",
            user_id: "User ID",
            enable_coin_magazines : "Coin Magazines",
            coin_magazine_amount : "Coin Magazine Amount",
            envelope_amount: "Envelope Amount"

        }
	});

	Globalize.ebos_translate = function(msg) {
		var translation = Globalize.formatMessage(msg);
		if (typeof translation === "undefined") {
			return "No Translation";
		} else {
			return translation;
		}
	};

	return Globalize
}));
