// ebos_public/public/js/filters.js
(function(exports) {

	exports.titleCase = function(string) {
		var retString = string.replace(/\w*/g, function(txt) {
			if (txt.substr(0,2).match(/[Mm][Cc]/)) {
				return 'Mc' + txt.charAt(2).toUpperCase() + txt.substr(3).toLowerCase();
			} else {
				return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
			}
		});

		return retString.replace(/([^\']*)\'S(\s|$)/, "$1's$2")
	};


	exports.commaNum = function(num) {
		return num.toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, "$1,");
	};

	exports.prettyCol = function(column_name) {
		switch (column_name) {
			case "case_count":
				return "Outer Packs";
			case "inner_pack_count":
				return "Inner Packs";
			case "loose_count":
				return "Loose Units";
			case "units_count":
				return "Total Units";
			case "date_created":
				return	"Last Saved";
			case "basis_list":
				return "List Type";
			case "status":
				return "Status";
		}
	};

	exports.parseFloat = function(string) {
		return parseFloat(string);
	};

	exports.realQuotes = function(charString) {
		charString = charString.replace('&quot;', '"', 'gi');
		return charString;
	};

	exports.htmlCharQuotes = function(charString) {
		charString = charString.replace('"', '&quot;', 'gi');
		return charString;
	};

	exports.jsonStringify = function(charString) {
		charString = JSON.stringify(charString);
		return charString;
	};

	exports.jsonParse = function(charString) {
		charString = JSON.parse(charString);
		return charString;
	};

	//TODO: deprecate to use prettyISODate
	exports.prettyDate = function(ISODate) {
		var ISOArray = ISODate.split("T");
		var dateArray = ISOArray[0].split('-');
		var timeArray = ISOArray[1].split(':');
		return dateArray[1] + '/' + dateArray[2] + '/' + dateArray[0] + " " + timeArray[0] + ":" + timeArray[1];
	};

	exports.prettyISODate = function(ISODate) {
		var ISOArray = ISODate.split("T");
		var dateArray = ISOArray[0].split('-');
		var timeArray = ISOArray[1].split(':');
		return dateArray[1] + '/' + dateArray[2] + '/' + dateArray[0] + " " + timeArray[0] + ":" + timeArray[1];
	};

	exports.prettyShortDate = function(shortDate) {
		var dateArray = shortDate.split('-');
		return dateArray[1] + '/' + dateArray[2] + '/' + dateArray[0];
	};


})(typeof exports === 'undefined'? this['filters']={}: exports);
